source/language/openpulse.rst

```diff
@@ -0,0 +1,702 @@
+
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single `numerically-controlled oscillator (NCO) <https://en.wikipedia.org/wiki/Numerically-controlled_oscillator>`_).
+
+Frame Construction
+~~~~~~~~~~~~~~~~~~
+
+Frames can be constructed using the ``newframe`` command e.g.
+
+.. code-block:: javascript
+
+  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)
+
+would construct a frame with a frequency of 5 GHz and a phase of 0.0. When
+instantiated, the frame time starts at 0. ``frame``\s can also be copied using the
+``copyframe`` command
+
+.. code-block:: javascript
+
+  frame driveframe1 = newframe(5e9, 0.0);
+  frame driveframe2 = copyframe(driveframe1);
+  driveframe2.phase = driveframe1.phase + pi/2;
+
+This will generate a :math:`\pi/2` phase incremented copy of ``driveframe1`` with the same
+frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase) tuple of a frame can be manipulated throughout program
+by referencing ``.frequency``, and ``.phase``. Operations must be
+appropriate for the respective type, ``float`` for frequency and ``angle`` for
+phase. Again, the exact precision of these calculations is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
+``driveframe``:
+
+.. code-block:: javascript
+
+   // Example 1: Shift phase of the "drive" frame by pi/4, to realize a virtual rz gate with angle -pi/4
+   cal {
+     driveframe.phase += pi/4;
+   }
+
+   // The following is an example only. Frames as arrays has not been agreed on.
+   // This conceptually must be compile-time arrays and treat qubits as indices
+   // which also has not been well-defined. We are exploring other solutions to
+   // the problem of mapping qubits to pulse-level resources.
+
+   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
+   cal {
+     array[frame, 8] rz_frames;
+     frame[0] = newframe(...);
+     // and so on
+   }
+
+   defcal rz(angle[20] theta) $q {
+     rz_frames[q].phase -= theta;
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: javascript
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Waveforms
+---------
+
+Waveforms are of type ``waveform`` and can either be:
+
+- An array of complex samples which define the points for the waveform envelope
+- An abstract mathematical function representing a waveform. This will later be
+  materialized into a list of complex samples, either by the compiler or the hardware
+  using the parameters provided to the ``extern`` defined pulse template.
```

Comment:
Maybe add a sentence based on offline discussion: like other `extern` functions, these will be compiled. But for static waveforms, the optimizing compiler should decide to execute this at compile time and load the waveform into memory once. For dynamic waveforms, the compiler just compiles and links this, to be executed at runtime.

Reply:
Great suggestion, added.